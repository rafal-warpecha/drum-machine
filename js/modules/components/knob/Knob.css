.container {
    position: relative;
    margin: 200px auto;
}

.input {
    position: absolute;
    top: 50%;
    left: 50%;
    width: 50%;
    height: 20px;
    margin-top: -10px;
    margin-left: -25%;
    padding: 0;
    outline: 0;
    border: 0;
    text-align: center;
    background-color: transparent;
    appearance: textfield;
}

.input::-webkit-outer-spin-button,
.input::-webkit-inner-spin-button {
    appearance: none;
}
