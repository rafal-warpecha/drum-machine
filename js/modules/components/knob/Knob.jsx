import React, { Component } from 'react';
import PropTypes from 'prop-types';
import KnobCanvas from './KnobCanvas';
import KnobInput from './KnobInput';
import styles from './Knob.css';

const { PI } = Math;

export default class Knob extends Component {
    static propTypes = {
        onChange: PropTypes.func.isRequired,
        min: PropTypes.number,
        max: PropTypes.number,
        value: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ]),
        size: PropTypes.number,
        thickness: PropTypes.number,
        angle: PropTypes.number,
        offsetAngle: PropTypes.number,
        fgColor: PropTypes.string,
        bgColor: PropTypes.string
    };

    static defaultProps = {
        min: 0,
        max: 100,
        value: 0,
        size: 70,
        thickness: 14,
        angle: 250,
        offsetAngle: -125,
        fgColor: '#ff0000',
        bgColor: '#aaaaaa'
    };

    static degreesToRadians(value) {
        return (value * PI) / 180;
    }

    static calculateStartAngle(offset) {
        return (1.5 * PI) + Knob.degreesToRadians(offset);
    }

    static calculateEndAngle(angle, offset) {
        return (1.5 * PI) + Knob.degreesToRadians(angle) + Knob.degreesToRadians(offset);
    }

    render() {
        const {
            onChange,
            value,
            min,
            max,
            size,
            thickness,
            angle,
            offsetAngle,
            fgColor,
            bgColor
        } = this.props;

        const sizeStyles = {
            width: `${size}px`,
            height: `${size}px`
        };

        return (
            <div className={styles.container} style={sizeStyles}>
                <KnobCanvas
                    onChange={onChange}
                    value={value}
                    min={min}
                    max={max}
                    size={size}
                    thickness={thickness}
                    fgColor={fgColor}
                    bgColor={bgColor}
                    angle={Knob.degreesToRadians(angle)}
                    startAngle={Knob.calculateStartAngle(offsetAngle)}
                    endAngle={Knob.calculateEndAngle(angle, offsetAngle)}
                    offsetAngle={Knob.degreesToRadians(offsetAngle)}
                />
                <KnobInput
                    onChange={onChange}
                    value={value}
                    min={min}
                    max={max}
                />
            </div>
        );
    }
}
