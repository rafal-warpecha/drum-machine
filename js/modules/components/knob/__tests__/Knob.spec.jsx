import React from 'react';
import { shallow } from 'enzyme';
import Knob from './../Knob';

describe('Knob component', () => {
    it('should match snapshot', () => {
        const wrapper = shallow(<Knob onChange={() => {}} />);

        expect(wrapper).toMatchSnapshot();
    });

    it('should convert degrees to radians', () => {
        expect(Knob.degreesToRadians(180)).toBeCloseTo(3.14, 2);
    });

    it('should calculate start angle', () => {
        expect(Knob.calculateStartAngle(20)).toBeCloseTo(5.1, 1);
    });

    it('should calculate end angle', () => {
        expect(Knob.calculateEndAngle(180, 20)).toBeCloseTo(8.2, 1);
    });
});
