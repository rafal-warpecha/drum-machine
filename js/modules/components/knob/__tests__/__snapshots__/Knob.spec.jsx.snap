// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Knob component should match snapshot 1`] = `
<div
  className="container"
  style={
    Object {
      "height": "70px",
      "width": "70px",
    }
  }
>
  <KnobCanvas
    angle={4.363323129985823}
    bgColor="#aaaaaa"
    endAngle={6.894050545377603}
    fgColor="#ff0000"
    max={100}
    min={0}
    offsetAngle={-2.1816615649929116}
    onChange={[Function]}
    size={70}
    startAngle={2.530727415391778}
    thickness={14}
    value={0}
  />
  <KnobInput
    max={100}
    min={0}
    onChange={[Function]}
    value={0}
  />
</div>
`;
